### PR TITLE
feat: Add ToolsAgent for models with tool-calling support

### PR DESCRIPTION
**Pull request by Heinrich-vanNieuwenhuizen**
### Description

This pull request introduces a new `ToolsAgent` that simplifies the integration of tools into agents powered by language models such as `Ollama` and `OpenAI`. The agent supports tool-driven workflows, enhancing the flexibility of LangChain agents.  

### Dependencies

- Added dependencies for `langchain_community`, `langchain_openai`, and `langchain_ollama`.

### Testing

- Unit tests for the `ToolsAgent` have been added, covering various tools and scenarios to ensure robust functionality.
- Example tests demonstrate integration with the `CalculatorTool` and memory.

@davidmigloz & @Ganeshsivakumar

### CODE REVIEW

File: examples/docs_examples/bin/modules/model_io/models/chat_models/integrations/ollama.dart [hunk 1]
```diff
@@ -109,7 +109,7 @@ Future<void> _chatOllamaToolCalling() async {
 
   final chatModel = ChatOllama(
     defaultOptions: const ChatOllamaOptions(
-      model: 'llama3.1',
+      model: 'llama3-groq-tool-use',
```
Comment (davidmigloz) [inline examples/docs_examples/bin/modules/model_io/models/chat_models/integrations/ollama.dart:112]: Actually, for zero-shot prompting `llama3.1` tool calling performs better than `llama3-groq-tool-use` per my experiments. It's only for multi-turn conversations when it doesn't get it right (which is a [known limitation](https://arc.net/l/quote/eqgeluxc) of `llama3.1`).

In this example, `llama3.1` is able to call the two tools in parallel, whereas `llama3-groq-tool-use` only calls one.

File: docs/modules/agents/agent_types/tool_agent.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
+# Tools Agent
+
+The Tools Agent is a flexible and powerful agent that can work with various language models and tools. It's designed to interpret the language model's output and decide when to call specific tools based on that output. This agent is particularly useful when you want to use models that don't have built-in function-calling capabilities or when you want more control over the tool-calling process.
+
+The Tools Agent can be used with a wide range of models, including those from OpenAI, Anthropic, Ollama, and others. It interprets the model's text output to determine when and how to use tools.
+
+> **Note**: Can be used with any chat model, not just those with built-in function-calling capabilities.
```
Comment (davidmigloz) [inline docs/modules/agents/agent_types/tool_agent.md:7]: I don't think this use case is worth supporting. Almost all SOTA models already support tool calling, and all Ollama models that are powerful enough to use tools have also been updated to use the tools API.

If a user really wants to use a model that doesn't support tool calling, they can simply modify the default prompt template of the ToolsAgent.

**Review by davidmigloz — COMMENTED** [review]
Hey @Heinrich-vanNieuwenhuizen,

Thank you very much for completing the PR 🙂

I've made a commit processing all the comments I posted. Mainly:
- Removing the custom tool-calling prompt, as this is already handled by the providers. The goal of this agent is to support models that support the tool calling API. If the model doesn't support it, the user can still provide a custom prompt tailored to the specific model.
- Revert the examples to `llama3.1`, as it's a more powerful model than `llama3-groq-tool-use` for most use cases

I'll deprecate the `OpenAIToolsAgent` in a separate PR and point users to this agent instead.

File: packages/langchain/pubspec.yaml [hunk 3]
```diff
@@ -25,3 +25,7 @@ dependencies:
 
 dev_dependencies:
   test: ^1.25.8
+  langchain_community:
+  langchain_openai:
+  langchain_ollama:
+    path: ../langchain_ollama
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/langchain/pubspec.yaml]: fixed

File: packages/langchain/lib/src/agents/tools.dart [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        llm: llm,
+        llmOptions: llm.defaultOptions,
+        prompt: createPrompt(
+          // systemChatMessage: toolPrompt!,
+          systemChatMessage: SystemChatMessagePromptTemplate(
+            prompt: PromptTemplate(
+              inputVariables: const {},
+              template: '''
+              ${systemChatMessage.prompt.template}\n\n
+                  'You have access to these tools: ${llm.defaultOptions.tools!.map((t) => t.name).join(', ')}'
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/langchain/lib/src/agents/tools.dart]: removed

File: examples/docs_examples/bin/modules/model_io/models/chat_models/integrations/ollama.dart [hunk 5]
```diff
@@ -27,7 +27,7 @@ Future<void> _chatOllama() async {
...
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply examples/docs_examples/bin/modules/model_io/models/chat_models/integrations/ollama.dart]: reverted

File: examples/docs_examples/bin/modules/model_io/models/chat_models/integrations/ollama.dart (hunk 1 above)
Comment (davidmigloz) [reply examples/docs_examples/bin/modules/model_io/models/chat_models/integrations/ollama.dart]: reverted

File: docs/modules/agents/agent_types/tool_agent.md (hunk 2 above)
Comment (davidmigloz) [reply docs/modules/agents/agent_types/tool_agent.md]: Removed

**Comment by davidmigloz** [comment]
@Heinrich-vanNieuwenhuizen Let me know if you have any comments on my changes, otherwise I'll merge it!

**Review by davidmigloz — APPROVED** [review]

File: packages/langchain/lib/src/agents/tools.dart [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+  ///
+  /// - [llm] - The model to use for the agent.
+  /// - [llmOptions] - The options to use for the model, must include the tools
+  /// - [memory] - The memory to use for the agent.
+  /// - [systemChatMessage] message to use as the system message that will be
+  ///   the first in the prompt. Default: "You are a helpful AI assistant".
+  /// - [extraPromptMessages] prompt messages that will be placed between the
+  ///   system message and the input from the agent.
+  factory ToolsAgent.fromLLMAndTools({
+    required final BaseChatModel llm,
```
[the remark this answers is not in the document]
Comment (Heinrich-vanNieuwenhuizen) [reply packages/langchain/lib/src/agents/tools.dart:127]: Sure, I'll change it.

**Comment by Heinrich-vanNieuwenhuizen** [comment]
> * `a3.1`, as it's a more powerful model than `llama3-groq-tool-use` for most use ca

Hi @davidmigloz , do your tests pass with llama3.1 mine only works with llama3-groq-tool-use. I 

I had mixed results without "Removing the custom tool-calling prompt, as the providers already handle this. "

File: packages/langchain/lib/src/agents/tools.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/langchain/lib/src/agents/tools.dart:127]: I updated it as well in my commit

**Comment by davidmigloz** [comment]
> do your tests pass with `llama3.1` mine only works with `llama3-groq-tool-use`?

I left `llama3-groq-tool-use` for the `tools_test.dart`. Although with `llama3.1` all the tests pass except the "Test ToolsAgent with messages memory" test, which is a [known limitation](https://arc.net/l/quote/eqgeluxc) of `llama3.1`:

> We recommend using Llama 70B-instruct or Llama 405B-instruct for applications that combine conversation and tool calling. Llama 8B-Instruct can not reliably maintain a conversation alongside tool calling definitions. It can be used for zero-shot tool calling, but tool instructions should be removed for regular conversations between the model and the user.

> I had mixed results without the custom tool-calling prompt

If we add that custom prompt, the actual prompt that arrives to the model will contain two tools definitions. The one from our custom prompt plus [the tool calling prompt that Ollama adds](https://arc.net/l/quote/pzrgpwgj) (or the other providers). So that will probably confuse the model (and add many unnecessary input tokens). The custom prompt should only be necessary for models that don't support tool calling. In that case, the user can provide a custom system prompt:
```dart
final agent = ToolsAgent.fromLLMAndTools(
  llm: llm,
  systemChatMessage: <their custom prompt>,
);
```

File: packages/langchain/lib/src/agents/tools.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (Heinrich-vanNieuwenhuizen) [reply packages/langchain/lib/src/agents/tools.dart:127]: Okay, I saw you already did it

**Comment by Heinrich-vanNieuwenhuizen** [comment]
> @Heinrich-vanNieuwenhuizen Let me know if you have any comments on my changes, otherwise I'll merge it!

Okay looks good you can merge

**Comment by davidmigloz** [comment]
Merged! Thanks again for your contribution @Heinrich-vanNieuwenhuizen 🙌
I'll release it this week.

**Comment by davidmigloz** [comment]
I've tested it with [mistral-nemo](https://ollama.com/library/mistral-nemo), and it's also working pretty well.

**Pull request merged**